### PR TITLE
ENH: Add Loss Outliers Analysis callback

### PR DESCRIPTION
**Pull request by kenza-bouzid**
Add LossAnalysisCallback to plot loss heatmaps + scatter plots and save loss ranks and their statistics.

### CODE REVIEW

File: hi-ml-cpath/Makefile [hunk 1]
```diff
@@ -78,7 +78,7 @@ pytest_coverage:
 	pytest --cov=health_cpath --cov SSL --cov-branch --cov-report=html --cov-report=xml --cov-report=term-missing --cov-config=.coveragerc
 
 SSL_CKPT_RUN_ID_CRCK := CRCK_SimCLR_1655731022_85790606
-SRC_CKPT_RUN_ID_CRCK := TcgaCrckSSLMIL_1659618104_10061e60
```
Comment (kenza-bouzid) [inline hi-ml-cpath/Makefile:81]: Update checkpoint to latest changes to fix issue with PL save_hyperparameter

File: .github/workflows/cpath-pr.yml [hunk 2]
```diff
@@ -215,6 +215,22 @@ jobs:
           cd ${{ env.folder }}
           make smoke_test_crck_flexible_finetuning_aml
 
+  smoke_test_crck_loss_analysis_aml:
```
Comment (kenza-bouzid) [inline .github/workflows/cpath-pr.yml:218]:
```suggestion
  smoke_test_crck_loss_analysis:
```

File: .github/workflows/cpath-pr.yml [hunk 3]
```diff
@@ -224,7 +240,8 @@ jobs:
       smoke_test_slidespandaimagenetmil,
       smoke_test_tilespandaimagenetmil,
       smoke_test_crck_simclr_aml,
-      smoke_test_crck_flexible_finetuning_aml
+      smoke_test_crck_flexible_finetuning_aml,
+      smoke_test_crck_loss_analysis_aml,
```
Comment (kenza-bouzid) [inline .github/workflows/cpath-pr.yml:244]:
```suggestion
      smoke_test_crck_loss_analysis,
```

File: hi-ml-cpath/testhisto/testhisto/utils/test_callbacks.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    # Loss cache is flushed after each epoch
+    _assert_loss_cache_contains_n_elements(loss_callback.loss_cache, 0)
+
+    if rank > 0:
+        time.sleep(10)  # Wait for the rank 0 to save the loss cache in a csv file
+
+    loss_cache_path = loss_callback.cache_folder / LOSS_VALUES_FILENAME.format(loss_callback.zfill_epoch(current_epoch))
+    assert loss_callback.cache_folder.exists()
+    assert loss_cache_path.exists()
+    assert loss_cache_path in loss_callback.cache_folder.iterdir()
```
Comment (dccastro) [inline hi-ml-cpath/testhisto/testhisto/utils/test_callbacks.py:165]: [Minor] What't the purpose of this check? Do you mean `local_cache_path.parent == loss_callback.cache_folder`?

File: hi-ml-cpath/testhisto/testhisto/utils/test_callbacks.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    # check plot_slides_loss_scatter outputs
+    assert (loss_callback.scatter_folder / SCATTER_PLOT_FILENAME.format(HIGHEST)).exists()
+    assert (loss_callback.scatter_folder / SCATTER_PLOT_FILENAME.format(LOWEST)).exists()
+
+    # check plot_loss_heatmap_for_slides_of_epoch outputs
+    for epoch in range(max_epochs):
+        filename = HEATMAP_PLOT_FILENAME.format(loss_callback.zfill_epoch(epoch), HIGHEST)
+        assert (loss_callback.heatmap_folder / filename).exists()
+        filename = HEATMAP_PLOT_FILENAME.format(loss_callback.zfill_epoch(epoch), LOWEST)
+        assert (loss_callback.heatmap_folder / filename).exists()
```
Comment (dccastro) [inline hi-ml-cpath/testhisto/testhisto/utils/test_callbacks.py:212]: Consider creating convenience methods to generate all these file paths (especially the ones with variables), for readability and to ensure they're always formatted the same way.

**Review by dccastro — CHANGES_REQUESTED** [review]

File: hi-ml-cpath/src/health_cpath/utils/callbacks.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    @property
+    def heatmap_folder(self) -> Path:
+        return self.outputs_folder / "loss_heatmap"
+
+    @property
+    def rank_folder(self) -> Path:
+        return self.outputs_folder / "loss_ranks"
+
+    @property
+    def exception_folder(self) -> Path:
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/callbacks.py:144]: Anomalies is definitely better. Will rename it.

File: hi-ml-cpath/src/health_cpath/utils/callbacks.py [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+        return first_time or (
+            current_epoch > self.patience and (current_epoch - self.patience) % self.epochs_interval == 0
+        )
+
+    def save_loss_cache(self, current_epoch: int) -> None:
+        """Saves the loss cache to a csv file"""
+        loss_cache_df = pd.DataFrame(self.loss_cache)
+        loss_cache_df = loss_cache_df.sort_values(by=ResultsKey.LOSS, ascending=False)
+        loss_cache_df.to_csv(
+            self.cache_folder / LOSS_VALUES_FILENAME.format(self.zfill_epoch(current_epoch)), index=False
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/callbacks.py:182]: Good point, done!

File: hi-ml-cpath/src/health_cpath/utils/callbacks.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+    def select_loss_slides_across_epochs(self, high: bool = True) -> Tuple[np.ndarray, np.ndarray]:
+        """Selects the slides with the highest/lowest loss values across epochs
+
+        :param high: If True, selects the slides with the highest loss values, else selects the slides with the lowest
+        loss values.
+        """
+        slides = []
+        slides_loss = []
+        for epoch in self.epochs_range:
+            loss_cache = pd.read_csv(self.cache_folder / LOSS_VALUES_FILENAME.format(self.zfill_epoch(epoch)))
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/callbacks.py]: Good point.

File: hi-ml-cpath/src/health_cpath/utils/callbacks.py [hunk 9]
```diff
@@ -200,67 +196,92 @@ def gather_loss_cache(self, rank: int) -> None:
...
 
         slides_loss_values: LossDictType = {slide_id: [] for slide_id in slides}
         for epoch in self.epochs_range:
-            loss_cache = pd.read_csv(self.cache_folder / LOSS_VALUES_FILENAME.format(self.zfill_epoch(epoch)))
-            loss_cache.set_index(ResultsKey.SLIDE_ID, inplace=True)
+            loss_cache = self.read_loss_cache(epoch, idx_col=ResultsKey.SLIDE_ID)
             slides_loss_values = {
                 slide_id: slides_loss_values[slide_id] + [loss_cache.loc[slide_id, ResultsKey.LOSS]]
                 for slide_id in slides
             }
```
Comment (dccastro) [inline hi-ml-cpath/src/health_cpath/utils/callbacks.py:255]: On second thought, this could be a clearer implementation, without needing to recreate the dict and lists again for every epoch:
```python
for slide_id in slides:
    slide_loss_values[slide_id].append(loss_cache.loc[slide_id, ResultsKey.LOSS])
```

File: hi-ml-cpath/src/health_cpath/utils/callbacks.py [hunk 10]
```diff
@@ -200,67 +196,92 @@ def gather_loss_cache(self, rank: int) -> None:
...
+        """Selects the slides with the highest/lowest loss values across epochs
+
+        :param high: If True, selects the slides with the highest loss values, else selects the slides with the lowest
+        loss values.
+        """
+        slides = []
+        slides_loss = []
+        for epoch in self.epochs_range:
+            epoch_slides = self.select_slides_for_epoch(epoch, high, self.num_slides_scatter)
+            epoch_slides_loss = self.select_slides_losses_for_epoch(epoch, high, self.num_slides_scatter)
```
Comment (dccastro) [inline hi-ml-cpath/src/health_cpath/utils/callbacks.py:268]: This will call `read_loss_cache()` twice for each `epoch` (once for slide IDs and once for losses), no? Do you expect this to have any impact on performance?

File: hi-ml-cpath/testhisto/testhisto/utils/test_callbacks.py (hunk 5 above)
Comment (dccastro) [reply hi-ml-cpath/testhisto/testhisto/utils/test_callbacks.py]: In my opinion, the API would be cleaner with a separate method to get the path for each of these files, so the user doesn't have to remember which folder and parameters need to be provided for each of them. E.g.
- `get_all_epochs_file()`
- `get_loss_values_file(epoch)`
- `get_heatmap_plot_file(epoch, order)`
- etc.

What do you think?

File: hi-ml-cpath/src/health_cpath/utils/callbacks.py [hunk 11]
```diff
@@ -141,48 +142,43 @@ def rank_folder(self) -> Path:
...
 
-    def reset_loss_cache(self) -> LossDictType:
+    def reset_loss_cache(self) -> LossCacheDictType:
         keys = [ResultsKey.LOSS, ResultsKey.SLIDE_ID]
         if self.save_tile_ids:
             keys.append(ResultsKey.TILE_ID)
         return {key: [] for key in keys}
 
-    def is_time_to_cache_loss_values(self, current_epoch: int) -> bool:
+    def get_filename(self, root_folder: Path, filename: str, epoch: int, order: Optional[str] = None) -> Path:
```
Comment (dccastro) [inline hi-ml-cpath/src/health_cpath/utils/callbacks.py:165]: Using `filename` in the name a method that returns a full path can be confusing, as that usually refers to the last part of a path.

**Review by dccastro — APPROVED** [review]
Great work, @kenza-bouzid! Just a few follow-up suggestions.
